Ignore unknown es query params

hmstepanek opened the pull request:
Add a schema to ignore and remove all unknown search query params when running a GET /api/search request and running a search on activity pages. This is a fix for https://github.com/hypothesis/product-backlog/issues/741.

hmstepanek commented on h/schemas/annotation.py:
```diff
@@ -311,3 +311,86 @@ def _target_selectors(targets):
+        'required': [],
+    }
+
+    def validate(self, data):
```
I didn't make a different validator but rather overrode the validate method. This seemed like the simplest approach but I'm open to other ideas. The validator we are using already ignores the unknown properties but it passes them through rather than removing them so all we have to do after the validator runs is remove them.

hmstepanek commented on tests/h/schemas/annotation_test.py:
```diff
@@ -608,6 +613,70 @@ def test_it_passes_existing_target_uri_to_document_metas_from_data(
+
+        assert params == expected_params
+
+    def test_raises_if_invalid_sorting_order(self, schema):
```
I'm thinking if we do this we may need to handle this "nicely" in the client and activity pages?

hmstepanek commented on h/schemas/annotation.py:
```diff
@@ -311,3 +311,88 @@ def _target_selectors(targets):
+    """Validates a JSON payload for an elasticsearch query via API"""
+    schema = {
+        # 'type': 'object',
+        'properties': {
```
We need to have a discussion/come to an agreement on these/aka which fields we want to make query-able. I have eliminated some I felt we pretty obviously didn't want to expose in a particular commit (of course there may be some discussion on those too).

hmstepanek replied:
Once we agree on these then I will add them to the api docs.

hmstepanek commented on h/activity/query.py:
```diff
@@ -42,7 +43,8 @@ def extract(request, parse=parser.parse):
     """
 
-    q = parse(request.params.get('q', ''))
+    schema = SearchParamsSchema()
```
Lyza asked why: Because this function takes in a request obj and pulls stuff off the request obj besides the params I think it's easier to leave this function's api as is and just put the schema call inside here. The alternative is breaking out all the things this function needs off the request obj and passing them in as separate args.

hmstepanek commented:
Considering the possibility of switching libraries: http://docs.python-cerberus.org/en/stable/normalization-rules.html

hmstepanek commented:
Lyza asked me what happens when you pass in for example; order:no_exist (aka an invalid sort order):
This would fail with the following error:
```
elasticsearch.exceptions.RequestError: TransportError(400, 'parsing_exception', '[field_sort] failed to parse field [order]')
```

hmstepanek commented on h/schemas/annotation.py:
```diff
@@ -311,3 +311,71 @@ def _target_selectors(targets):
+                'type': 'string',
+            },
+            'limit': {
+                'type': 'string',
```
We already discussed this but just for doc purposes: No, there is not built in check to verify a string can be converted to an int. I think we could write our own thing but I think it's simpler just to let the lower level code handle it like we are currently doing.

robertknight commented on h/schemas/base.py:
```diff
@@ -12,6 +12,9 @@
 from pyramid.session import check_csrf_token
 
 
+PROPERTIES = "properties"
```
Any particular reason this is a constant? I thought you preferred to avoid constants whose names are basically their values?

hmstepanek replied:
My original thought was if we used a different schema validator it might change but I don't think that's gonna happen anymore so I'll remove it.

hmstepanek commented on h/schemas/activity.py:
```diff
@@ -0,0 +1,69 @@
+# -*- coding: utf-8 -*-
+"""Classes for validating data passed to activity pages."""
```
It seems like this is something we should cleanup? I'm not gong to try to tackle it in this PR but I think perhaps making a simple parser that simply converts things into a dict format and passes them to the jsonschema validator might make things a little cleaner here. I think this is a small piece of cleaning up activity pages but every little bit helps. What do you think?

hmstepanek commented on docs/_extra/api-reference/hypothesis.yaml:
```diff
@@ -305,10 +305,11 @@ paths:
         - name: order
           in: query
-          description: The order in which the results should be sorted.
+          description: The direction of sort.
```
I felt like this description was more concise and hinted at the fact that there are specific directions but if you don't like it I can change it back.

robertknight replied:
I found the previous description marginally clearer myself.

hmstepanek replied:
Changed it back.

robertknight commented on tests/h/schemas/base_test.py:
```diff
@@ -87,6 +116,90 @@ def test_it_returns_all_errors_in_message(self):
+        (
+            # data
+            {"foo": "string",
+             "bar": {
```
"bar" is defined as being an integer in the schema. `remove_unused_properties` should never be called with this input because the data should have been validated beforehand and this input would fail validation.

Additionally I found it surprising that `bar` is expected to be stripped from the validated data since `bar` is a valid field in the schema.

I discovered this when trying to figure out if the `remove_unused_properties` implementation could be simplified.

hmstepanek replied:
See https://github.com/hypothesis/h/pull/5224/files#r212778211.

robertknight commented on docs/_extra/api-reference/hypothesis.yaml:
```diff
@@ -305,10 +305,11 @@ paths:
           description: The field by which annotations should be sorted.
           required: false
           type: string
+          enum: [created, updated, group, id, tags_raw, raw_user, uri]
```
See my comments about `tags_raw` and `raw_user` in `h/schemas/annotation.py`.

robertknight commented on h/schemas/annotation.py:
```diff
@@ -311,3 +311,82 @@ def _target_selectors(targets):
+
+
+class SearchParamsSchema(JSONSchema):
+    """Validates a JSON payload for an elasticsearch query via API"""
```
This description needs updating. The schema validates query parameters to an `/api/search` request, not the JSON query that is sent to Elasticsearch - that's an implementation detail of the search API.

robertknight commented on h/schemas/base.py:
```diff
@@ -73,6 +73,43 @@ def validate(self, data):
+        extra_props = set(data_d.keys()) - set(schema_d.keys())
+        remove_props(data_d, extra_props)
+
+        extra_props = []
```
For our current use case do we actually have any need to deal with nested fields? We're already leaving out handling for arrays amongst other things so we might as well leave that out if we don't need it.

hmstepanek replied:
Lyza wanted to make it work with nested fields. I'm also in agreement with that but yes, we are leaving out arrays, and other property fields. I have noted this in the description now. Judging by your reaction to this dropping a property I will treat the property that incorrectly contains nested properties as an invalid type and not drop it from the data. I will however, leave the if "properties" key check in place so that if someone runs this without validating it, it won't fail with an unhelpful key error.

robertknight commented on tests/h/schemas/base_test.py:
```diff
@@ -87,6 +116,90 @@ def test_it_returns_all_errors_in_message(self):
+            # data
+            MultiDict([("foo", "string"),
+                       ("foo", "str"),
+                       ("baz", {"foo": "string"}),
```
Is it actually possible to end up with a non-string value in `request.params`?

hmstepanek replied:
No. But again-the intention here was to make it generic enough that other schemas could potentially use this function.

robertknight commented on tests/h/schemas/base_test.py:
```diff
@@ -87,6 +116,90 @@ def test_it_returns_all_errors_in_message(self):
+                       ("baz", {"foo": "string"})]),
+
+    )])
+    def test_it_removes_unknown_properities_from_data(self, data, expected):
```
Typo "properities".

hmstepanek replied:
There are certain words that I kept swizzling letters on when I was writing this-I was having the worst time with "unknown" I kept leaving out "n"s so I'd type "uknown" or "unkown". Typing is hard. lol

hmstepanek commented on docs/_extra/api-reference/hypothesis.yaml:
```diff
@@ -342,10 +343,41 @@ paths:
+          type: string
+        - name: quote
+          in: query
+          description: Limit the results to annotations that contain this text in their textual body.
```
I don't know how I got these backwards I even looked them up in the model to make sure I got them right-I must have just swizzled the name at some point.  Every time I have to look them up cause I forget which is which.

hmstepanek commented on docs/_extra/api-reference/hypothesis.yaml:
```diff
@@ -342,10 +343,41 @@ paths:
+          type: string
+        - name: references
+          in: query
+          description: A list of parent annotation ids by which resulting annotations are
```
Maybe I'm misunderstanding what references is but I thought it was a list of parent ids that the annotation was replying to so that description you have there doesn't make sense. I hear what you are saying though-about the description being unclear and it looks like no you can't pass a list, only a single value.

hmstepanek commented on h/schemas/annotation.py:
```diff
@@ -311,3 +311,82 @@ def _target_selectors(targets):
+            },
+            'sort': {
+                'type': 'string',
+                'enum': ["created", "updated", "group", "id", "tags_raw", "raw_user", "uri"],
```
No it doesn't work so I'll remove it. To sort by user es requires the field be non-analyzed so we must use the user_raw field but it can be exposed as user and translated to user_raw before querying.

robertknight commented:
After some investigation I think that colander will be a better tool to validate query string parameters than JSON schema, since the input is much more similar to HTML form data (a flat list of string key / value pairs). What I would suggest doing from here is:

- Merging the helper added in https://github.com/hypothesis/h/pull/5247 (see tests for example usage)
- Converting the JSON schema in this PR to a colander schema and using the validation helper in #5247 to process the query params.

What this will give us is:

- Range and type validation for numeric fields
- Value validation for enum fields (which is not working in this PR)

hmstepanek commented:
@robertknight I've updated this PR to use the validate_query_params util function. Please note I also put a limit on offset. If that's too much I can split it out into it's own PR but I would like to merge that first as it also requires an update to the schema and the docs that I'm already touching in this PR.

hmstepanek commented:
I also added a test to validate_query_params. I noticed we didn't really have coverage for that when I started working on the schema. Again-I can also split this into it's own PR if you would like.

robertknight commented on h/schemas/annotation.py:
```diff
@@ -311,3 +313,99 @@ def _target_selectors(targets):
+    )
+    limit = colander.SchemaNode(
+        colander.Integer(),
+        validator=colander.Range(min=0, max=LIMIT_MAX),
```
It is possible there are existing API users out there passing in illegal values for limit or offset, so even though we are not making a change to the limits, there is a small risk that it causes a problem. I think we should still merge this, but we should keep an eye out and be prepared to adjust quickly if necessary.

hmstepanek replied:
I feel safe about using a limit of 10000 cause right now that is our limit anyway. I feel less safe about decreasing it to 2000 without first seeing what limits are being used today. I think I'll leave it at 10000 for now and add an event in to record the value and then based on that info we can decrease it to a more reasonable number w/o breaking anyone.

robertknight commented on h/schemas/annotation.py:
```diff
@@ -311,3 +313,99 @@ def _target_selectors(targets):
+        colander.SchemaNode(colander.String()),
+        missing=colander.drop,
+        description="""Limit the results to annotations whose quote, tags,
+                       text or url fields contain this key word.""",
```
"key word" should be one word.

robertknight commented on h/schemas/annotation.py:
```diff
@@ -311,3 +313,99 @@ def _target_selectors(targets):
+    _separate_replies = colander.SchemaNode(
+        colander.Boolean(),
+        missing=False,
+        description="Return a seperate set of annotations and their replies.",
```
typo - s/seperate/*separate*. I've mixed this up before.

hmstepanek commented:
^ reverted because some of the changes for sorting got lost-i'll reopen with the sorting changes included.
